stellargraph/mapper/sequences.py

```diff
@@ -273,20 +272,12 @@ def __init__(self, sample_function, batch_size, walker, shuffle=True):
                 "({}) UnsupervisedSampler is required.".format(type(self).__name__)
             )
 
-        self.lock = threading.Lock()
         self.batch_size = batch_size
         self.walker = walker
-
-        # an estimate of the  upper bound on how many samples are generated in each epoch
-        self.data_size = (
-            2
-            * len(self.walker.nodes)
-            * self.walker.length
-            * self.walker.number_of_walks
-        )
-
-        # the generator method from the sampler with the batch-size from the link generator method
-        self._gen = self.walker.generator(self.batch_size)
+        self.shuffle = shuffle
+        self._batches = self._create_batches()
```

Comment:
It seems like putting this call here is converting this `__init__` constructor from being efficient and on-demand per the class's name (O(1) time and memory?) to doing all of the sampling and storing it in memory (i.e. O(size of sample)). This seems suboptimal. Can we avoid it?

Reply:
I guess one way would be putting in a check in `__getitem__` to see if batches exist, otherwise create them.

I'll double check if `fit_generator` is supposed to call any sort of pre-train method on the sequence

Reply:
Why not continue to create them on demand, doing the sampling for each batch in `__getitem__`?

Reply:
Hmmm.. I guess we could but then it would require similar sort of dictionary of samplers logic in `unsupervised_sampler` as well, since the order of `batch_num` is not guaranteed. Since this is just creating the IDs associated with each training example without features, I didn't think it was particularly worth doing that.

Reply:
> I guess we could but then it would require similar sort of dictionary of samplers logic in unsupervised_sampler as well

Right.

> Since this is just creating the IDs associated with each training example without features, I didn't think it was particularly worth doing that.

Good point, but I think the explosion in memory use is still potentially concerning: going from a number of pairs of nodes of size `O(batch_size)` (since only a single batch is buffered at a time) to `O(number_of_walks * nodes * length)`; this seems like it *could* approach or even exceed the size of the features if `number_of_walks * length` is a significant fraction of the feature dimension, although in practice, in our demos, it seems like that's at most `5 * 4`.

If we are going to do this, I think we should potentially consider renaming the class, since it's no longer on demand? (Possibly with some sort of nice deprecation warnings.) Either that, or land this as is but have future work to refactor/improve it.

What do you think?

Reply:
Hmm.. I'm not sure 🤷‍♂ I'm kind of leaning towards just keeping this "on-demand" by adding the batch num logic here too.

Reply:
Hmm actually I think doing each batch on demand is more difficult here than in the link generator... basically if `batch_size % (number_of_walks * (length - 1) * 2) != 0` then it gets kind of awkward (the `* 2` is for each positive context having a negative context)  For instance:

With 1 walk of length 3, i.e. 2 positive context pairs per node + 2 negative, we'd be generating 4 examples per node. If `batch_size` doesn't divide evenly into 4's, then we can't just assign nodes to batches, we actually have to assign individual context pairs to batches, in which case we'd basically end up with the same sort of memory footprint? 

To avoid the memory footprint I think we'd need all threads to share one "state" of context pairs being generated on the fly (similarly to how it was implemented previously where `next(generator)` was wrapped around a lock)... but I'm not sure that can be made reproducible unless we just force batches to be processed in order (i.e. `UnsupervisedSampler` keeps track of which `batch_num` to process next, and has to block other threads)

Reply:
Ah, great point.

Do the batches all have to be exactly the same size? If not, I think we can get down to `O(number_of_walks * length)` (or something) by having a dummy batch with any left overs every time we overfill a batch. That is, if it doesn't divide evenly, we end up generating something like 2&times; more batches: one full data and one with leftovers, and they have to be generated at the same time and cached until the other one is requested, e.g. if `batch_size = 3` in your example:

| batch | size |
|---|---:|
|0|3|
|1|1|
|2|3|
|3|1|
|...|...|

I guess in the worst case, this can result in caching half of the batches, if it requests 0, 2, 4, ... without requesting the others, but usually won't?

I'd be happy to leave this work for future, given it seems moderately complicated.

> UnsupervisedSampler keeps track of which batch_num to process next, and has to block other threads

I'd be concerned that this might deadlock, if for whatever reason, there was no thread trying to process the next batch in order (e.g. if it starts with `batch_num` 10, 20, 30, they'll all be waiting forever for 0, 1, ..., 9 to be finished).

Reply:
I've created #681 for now

Things *might* still work without ensuring all batches have the same size... but even if it does work, it feels like we'd breaching the contract of using Keras's API and I'd be concerned about running into more subtle problems, unless Keras explicitly says this is ok.

Reply:
```suggestion
        # FIXME(#681): all batches are created at once, so this is no longer "on demand"
        self._batches = self._create_batches()
```